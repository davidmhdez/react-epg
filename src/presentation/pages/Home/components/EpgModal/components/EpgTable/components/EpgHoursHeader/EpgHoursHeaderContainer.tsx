import React, {useMemo} from 'react';
import EpgHoursHeader from "./EpgHoursHeader.tsx";

type Props = {
    startDate: Date
    endDate: Date
}

const EpgHoursHeaderContainer: React.FC<Props> = ({ startDate, endDate }) => {

    const { width, initialOffset, timeIntervals } = useMemo(()=>{
        const adjustToNextHalfHour = (date: Date) => {
            const minutes = date.getMinutes();
            if (minutes !== 0 && minutes !== 30) {
                const adjustedMinutes = minutes < 30 ? 30 : 0;
                if (adjustedMinutes === 0) {
                    date.setHours(date.getHours() + 1);
                }
                date.setMinutes(adjustedMinutes, 0, 0);
            }
            return date;
        };

        const initialStartDate = new Date(startDate);
        const adjustedStartDate = adjustToNextHalfHour(new Date(startDate));

        // @ts-ignore
        const differenceInHours = (endDate - initialStartDate) / (1000 * 60 * 60);
        const width = differenceInHours * 300;


        const timeIntervals = [];
        const halfHourInMilliseconds = 30 * 60 * 1000;

        let currentTime = new Date(adjustedStartDate.getTime());
        while (currentTime <= endDate) {
            timeIntervals.push(new Date(currentTime));
            currentTime = new Date(currentTime.getTime() + halfHourInMilliseconds);
        }

        // @ts-ignore
        const initialOffset = (adjustedStartDate - initialStartDate) / (1000 * 60 * 30) * 150;

        return{
            width,
            initialOffset,
            timeIntervals
        }
    }, [startDate, endDate])


    return (
        <EpgHoursHeader
            initialOffset={initialOffset}
            width={width}
            timeIntervals={timeIntervals}
        />
    );
};

export default EpgHoursHeaderContainer;
