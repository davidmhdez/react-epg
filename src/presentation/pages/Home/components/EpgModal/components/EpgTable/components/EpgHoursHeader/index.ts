export {default} from "./EpgHoursHeaderContainer.tsx"