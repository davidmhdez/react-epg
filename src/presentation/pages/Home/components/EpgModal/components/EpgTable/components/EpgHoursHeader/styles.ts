import styled from "styled-components";

export const EpgHoursHeaderWrapper = styled.div`
    position: sticky;
    top: 50vh;
    background-color: #000000;
    z-index: 110;
    display: flex;
`;

export const EpgHeaderTitle = styled.p`
    width: 230px;
    color: #ffffff;
    font-size: 14px;
    text-align: center;
    margin: 0;
    font-weight: 900;
    position: sticky;
    left: 0;
    z-index: 120;
    padding: 10px 10px 10px 0;
    background-color: #000000;
`

export const EpgHoursList = styled.div<{$width: number}>`
    width: ${({ $width }) => $width}px;
    position: relative;
    padding: 10px 0;
    height: 37px;
`;

export const EpgHoursInterval = styled.div<{$index: number, $initialOffset: number}>`
    position: absolute;
    left: ${({$index, $initialOffset})=> $initialOffset + (300 / 2) * $index}px;
    transform: translateX(-50%);
    text-align: center;
    white-space: nowrap;
    color: #ffffff;
    
`