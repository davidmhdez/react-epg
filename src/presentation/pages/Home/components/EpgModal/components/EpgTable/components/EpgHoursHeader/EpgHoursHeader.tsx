import React from 'react';
import {formatTime} from "../../../../helpers.ts";
import {EpgHeaderTitle, EpgHoursHeaderWrapper, EpgHoursInterval, EpgHoursList} from "./styles.ts";

type Props = {
    width: number
    initialOffset: number
    timeIntervals: Date[]
}

const EpgHoursHeader: React.FC<Props> = ({ width, timeIntervals, initialOffset }) => {
    return (
        <EpgHoursHeaderWrapper>
            <EpgHeaderTitle>HOY</EpgHeaderTitle>
            <EpgHoursList $width={width}>
                {timeIntervals.map((time, index) => (
                    <EpgHoursInterval
                        key={index}
                        $index={index}
                        $initialOffset={initialOffset}
                    >
                        {formatTime(time)}
                    </EpgHoursInterval>
                ))}
            </EpgHoursList>
        </EpgHoursHeaderWrapper>
    );
};

export default EpgHoursHeader;