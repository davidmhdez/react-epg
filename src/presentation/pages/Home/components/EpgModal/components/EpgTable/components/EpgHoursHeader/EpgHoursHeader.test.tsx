import {render} from "@testing-library/react";
import EpgHoursHeader from "./EpgHoursHeaderContainer.tsx";

describe('EpgHoursHeader test', () => {
    test('should match with snapshot', () => {
        const fromDate = new Date(2024, 6, 25, 14, 0, 0);
        const toDate = new Date(2024, 6, 25, 23, 59, 59);
        const { container } = render(<EpgHoursHeader startDate={fromDate} endDate={toDate}/>)
        expect(container).toMatchSnapshot();
    })
})