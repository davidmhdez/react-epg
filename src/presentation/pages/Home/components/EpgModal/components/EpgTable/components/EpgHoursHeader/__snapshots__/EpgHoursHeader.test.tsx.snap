// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EpgHoursHeader test should match with snapshot 1`] = `
<div>
  <div
    class="sc-aYaIB fbKZa-d"
  >
    <p
      class="sc-gEvDqW gOaJWY"
    >
      HOY
    </p>
    <div
      class="sc-eqUzNf etwsRN"
    >
      <div
        class="sc-fqkwJk yBFZy"
      >
        14:00
      </div>
      <div
        class="sc-fqkwJk hkCXKG"
      >
        14:30
      </div>
      <div
        class="sc-fqkwJk haZKJR"
      >
        15:00
      </div>
      <div
        class="sc-fqkwJk fxoIHr"
      >
        15:30
      </div>
      <div
        class="sc-fqkwJk bmSnLo"
      >
        16:00
      </div>
      <div
        class="sc-fqkwJk hiuVqw"
      >
        16:30
      </div>
      <div
        class="sc-fqkwJk haMCRr"
      >
        17:00
      </div>
      <div
        class="sc-fqkwJk hLLrUO"
      >
        17:30
      </div>
      <div
        class="sc-fqkwJk eCocwF"
      >
        18:00
      </div>
      <div
        class="sc-fqkwJk hfNEzF"
      >
        18:30
      </div>
      <div
        class="sc-fqkwJk bOStyG"
      >
        19:00
      </div>
      <div
        class="sc-fqkwJk gOesHc"
      >
        19:30
      </div>
      <div
        class="sc-fqkwJk gUZdTP"
      >
        20:00
      </div>
      <div
        class="sc-fqkwJk hJnSAr"
      >
        20:30
      </div>
      <div
        class="sc-fqkwJk iRLhih"
      >
        21:00
      </div>
      <div
        class="sc-fqkwJk djsnVv"
      >
        21:30
      </div>
      <div
        class="sc-fqkwJk bQIeFg"
      >
        22:00
      </div>
      <div
        class="sc-fqkwJk gGpvpA"
      >
        22:30
      </div>
      <div
        class="sc-fqkwJk dVPsFL"
      >
        23:00
      </div>
      <div
        class="sc-fqkwJk jiqgzh"
      >
        23:30
      </div>
    </div>
  </div>
</div>
`;
